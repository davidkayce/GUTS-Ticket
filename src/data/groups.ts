export const groups = [
  {
    id: "+31611111111",
    seats: [
      { section: "main hall", row: "1", seat: "4" },
      { section: "main hall", row: "1", seat: "2" },
      { section: "main hall", row: "2", seat: "2" },
      { section: "main hall", row: "2", seat: "4" },
    ],
  },
  {
    id: "+31622222222",
    seats: [
      { section: "main hall", row: "3", seat: "4" },
      { section: "main hall", row: "3", seat: "2" }
    ],
  },
  {
    id: "+31633333333",
    seats: [
      { section: "vip region", row: "3", seat: "4" },
      { section: "vip region", row: "3", seat: "2" }
    ],
  }
]

