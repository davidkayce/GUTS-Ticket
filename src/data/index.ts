export * from './layout';
export * from './groups';
export * from './mocks';