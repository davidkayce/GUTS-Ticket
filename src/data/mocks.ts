export const mockGroup = [
  {
    id: "one",
    seats: [
      { section: "parlour", row: "3", seat: "4" },
      { section: "dining", row: "2", seat: "1" },
    ],
  },
  {
    id: "two",
    seats: [
      { section: "parlour", row: "2", seat: "4" },
      { section: "game room", row: "1", seat: "2" },
    ],
  },
];

export const mockLayout = {
  sections: [
    {
      name: "parlour",
      rows: [
        {
          row: "1",
          seats: [
            { seat: "1", rank: "rank1" },
            { seat: "3", rank: "rank1" },
            { seat: "4", rank: "rank1" },
            { seat: "2", rank: "rank1" },
          ],
        },
        {
          row: "2",
          seats: [
            { seat: "1", rank: "rank1" },
            { seat: "3", rank: "rank1" },
            { seat: "5", rank: "rank1" },
            { seat: "6", rank: "rank1" },
            { seat: "4", rank: "rank1" },
            { seat: "2", rank: "rank1" },
          ],
        },
        {
          row: "3",
          seats: [
            { seat: "1", rank: "rank1" },
            { seat: "3", rank: "rank1" },
            { seat: "5", rank: "rank1" },
            { seat: "6", rank: "rank2" },
            { seat: "4", rank: "rank2" },
            { seat: "2", rank: "rank2" },
          ],
        },
      ],
    },
    {
      name: "dining",
      rows: [
        {
          row: "1",
          seats: [
            { seat: "1", rank: "rank1" },
            { seat: "3", rank: "rank2" },
            { seat: "4", rank: "rank2" },
            { seat: "2", rank: "rank1" },
          ],
        },
        {
          row: "2",
          seats: [
            { seat: "1", rank: "rank1" },
            { seat: "3", rank: "rank2" },
            { seat: "5", rank: "rank2" },
            { seat: "6", rank: "rank3" },
            { seat: "4", rank: "rank3" },
            { seat: "2", rank: "rank1" },
          ],
        },
      ],
    },
    {
      name: "game room",
      rows: [
        {
          row: "1",
          seats: [
            { seat: "1", rank: "rank1" },
            { seat: "3", rank: "rank1" },
            { seat: "2", rank: "rank1" },
          ],
        },
      ],
    },
  ],
};

export const mockLayoutMatrix = {
  dining: [
    undefined,
    [
      { rank: "rank1" },
      { rank: "rank1" },
      { rank: "rank2" },
      { rank: "rank2" },
    ],
    [
      { rank: "rank1" },
      { rank: "rank1" },
      { rank: "rank2" },
      { rank: "rank3" },
      { rank: "rank2" },
      { rank: "rank3" },
    ],
  ],
  "game room": [
    undefined,
    [{ rank: "rank1" }, { rank: "rank1" }, { rank: "rank1" }],
  ],
  parlour: [
    undefined,
    [
      { rank: "rank1" },
      { rank: "rank1" },
      { rank: "rank1" },
      { rank: "rank1" },
    ],
    [
      { rank: "rank1" },
      { rank: "rank1" },
      { rank: "rank1" },
      { rank: "rank1" },
      { rank: "rank1" },
      { rank: "rank1" },
    ],
    [
      { rank: "rank1" },
      { rank: "rank2" },
      { rank: "rank1" },
      { rank: "rank2" },
      { rank: "rank1" },
      { rank: "rank2" },
    ],
  ],
};

export const mockAllocatedSeats = {
  dining: [
    undefined,
    [
      { rank: "rank1" },
      { rank: "rank1" },
      { rank: "rank2" },
      { rank: "rank2" },
    ],
    [
      { id: "one", rank: "rank1" },
      { rank: "rank1" },
      { rank: "rank2" },
      { rank: "rank3" },
      { rank: "rank2" },
      { rank: "rank3" },
    ],
  ],
  "game room": [
    undefined,
    [{ rank: "rank1" }, { id: "two", rank: "rank1" }, { rank: "rank1" }],
  ],
  parlour: [
    undefined,
    [
      { rank: "rank1" },
      { rank: "rank1" },
      { rank: "rank1" },
      { rank: "rank1" },
    ],
    [
      { rank: "rank1" },
      { rank: "rank1" },
      { rank: "rank1" },
      { id: "two", rank: "rank1" },
      { rank: "rank1" },
      { rank: "rank1" },
    ],
    [
      { rank: "rank1" },
      { rank: "rank2" },
      { rank: "rank1" },
      { id: "one", rank: "rank2" },
      { rank: "rank1" },
      { rank: "rank2" },
    ],
  ],
};
