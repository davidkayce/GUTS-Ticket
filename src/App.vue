<template>
  <section class="theater">
    <section>
      <div
        v-for="(value, name) in allocation"
        :key="name"
        class="theater-section"
      >
        <h3 class="section-name" :key="name">{{ name }}</h3>
        <table class="seats-container">
          <tr
            v-for="(row, rowIndex) in value"
            :key="rowIndex"
            class="seats-row"
          >
            <td v-for="(seat, seatIndex) in row" :key="seatIndex">
              <svg
                id="Layer_1"
                data-name="Layer 1"
                xmlns="http://www.w3.org/2000/svg"
                viewBox="0 0 100 100"
              >
                <path
                  :class="classifier(seat?.id, seat?.rank)"
                  d="M36,17.3H80.4a8.88,8.88,0,0,1,6.72-7.25A5.77,5.77,0,0,0,81.57,6H36a5.72,5.72,0,0,0-5.76,5.66A5.71,5.71,0,0,0,36,17.3Z"
                />
                <path
                  :class="classifier(seat?.id, seat?.rank)"
                  d="M80.29,82.79H36A5.66,5.66,0,1,0,36,94.1H81.47a6.13,6.13,0,0,0,5.44-3.41A8.77,8.77,0,0,1,80.29,82.79Z"
                />
                <path
                  :class="classifier(seat?.id, seat?.rank)"
                  d="M80.08,79.7V20.5H35.92A8.85,8.85,0,0,1,27.17,13h-18a4,4,0,0,0-4.06,4V82.79a4,4,0,0,0,4.06,3.95H27.28a8.65,8.65,0,0,1,8.75-7Z"
                />
                <path
                  :class="classifier(seat?.id, seat?.rank)"
                  d="M89.15,12.93a5.71,5.71,0,0,0-5.76,5.65V82.15a5.76,5.76,0,0,0,11.52,0V18.58A5.71,5.71,0,0,0,89.15,12.93Z"
                />
                <path
                  :class="classifier(seat?.id, seat?.rank)"
                  d="M90.21,9.94a8.93,8.93,0,0,0-8.74-7H36a8.94,8.94,0,0,0-8.75,6.93H9.15A7.22,7.22,0,0,0,2,17V82.79a7.06,7.06,0,0,0,7.15,7h18a8.85,8.85,0,0,0,8.75,7.26H81.47A8.91,8.91,0,0,0,90,90.9a8.81,8.81,0,0,0,8-8.75V18.58A8.84,8.84,0,0,0,90.21,9.94ZM36,6H81.57a5.77,5.77,0,0,1,5.55,4.06A8.88,8.88,0,0,0,80.4,17.3H36a5.71,5.71,0,0,1-5.76-5.65A5.72,5.72,0,0,1,36,6ZM27.28,86.74H9.15a4,4,0,0,1-4.06-3.95V17a4,4,0,0,1,4.06-4h18a8.85,8.85,0,0,0,8.75,7.47H80.08V79.7H36A8.65,8.65,0,0,0,27.28,86.74ZM81.47,94.1H36a5.66,5.66,0,1,1,0-11.31H80.29a8.77,8.77,0,0,0,6.62,7.9A6.13,6.13,0,0,1,81.47,94.1ZM94.91,82.15a5.76,5.76,0,0,1-11.52,0V18.58a5.76,5.76,0,0,1,11.52,0Z"
                />
              </svg>
            </td>
          </tr>
        </table>
      </div>
    </section>
    <article>
      <div class="card">
        <h3>Properties</h3>
        <div class="card-body">
          <ul class="list-group">
            <li class="list-group-item"><span class="group one" /> Group 1</li>
            <li class="list-group-item"><span class="group two" /> Group 2</li>
            <li class="list-group-item">
              <span class="group three" /> Group 3
            </li>
            <li class="list-group-item"><span class="rank one" /> Rank 1</li>
            <li class="list-group-item"><span class="rank two" /> Rank 2</li>
            <li class="list-group-item"><span class="rank three" /> Rank 3</li>
          </ul>
        </div>
      </div>
    </article>
  </section>
</template>

<script lang="ts">
import { ref, onMounted } from "vue";
import { defineComponent } from "vue";
import { layout, groups } from "./data";

export default defineComponent({
  name: "App",
  setup() {
    const buffer = {} as any;
    const allocation = ref<any>({});

    const idMap = {
      "+31611111111": "group1",
      "+31622222222": "group2",
      "+31633333333": "group3",
    };

    const drawSeats = () => {
      for (let i = 0, section; (section = layout.sections[i++]); ) {
        buffer[section.name] = [];
        for (let j = 0, row; (row = section.rows[j++]); ) {
          buffer[section.name][j] = [];
          for (let k = 0, seat; (seat = row.seats[k++]); ) {
            buffer[section.name][j][+seat.seat -1] = { rank: seat.rank };
          }
        }
      }
    };
    
    const allocateSeats = () => {
      for (let a = 0, group; (group = groups[a++]); ) {
        for (let b = 0, seat; (seat = group.seats[b++]); ) {
          buffer[seat.section][+seat.row][+seat.seat - 1]['id'] = group.id;
        }
      }
    };

    const classifier = (id: string, rank: number): string => {
      if (!id && !rank) return "seat";
      return `seat ${idMap[id] || ""} ${rank}`;
    };

    onMounted(() => {
      drawSeats();
      allocateSeats();
      allocation.value = buffer;
    });

    return {
      allocation,
      classifier,
    };
  },
});
</script>

<style lang="scss">
@import "./styles/main.scss";
</style>
