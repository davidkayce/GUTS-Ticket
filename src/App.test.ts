import { mount } from "@vue/test-utils";
import { mockGroup, mockLayout, mockLayoutMatrix, mockAllocatedSeats } from "./data";
import App from "./App.vue";


describe("Theater application", () => {
  it("should render the app properly", () => {
    const wrapper = mount(App, { props: null });
    expect(wrapper.exists()).toBe(true);
    expect(wrapper.find("h3").text()).toEqual("Properties");
  });

  test("should render content correctly", () => {
    const wrapper = mount(App);
    expect(wrapper.vm.$el).toMatchSnapshot();
  });

  test("draw seats function should generate proper matrix of layout", () => {
    const wrapper = mount(App);
    expect(wrapper.vm.drawSeats(mockLayout)).toEqual(mockLayoutMatrix);
  });

  test("allocate seats function should allocate seats according to groups", () => {
    const wrapper = mount(App);
    expect(wrapper.vm.allocateSeats(mockGroup, mockLayoutMatrix)).toEqual(mockAllocatedSeats);
  });

  test("the seats should be drawn properly", () => {
    const wrapper = mount(App);
    let totalRows = 0;
    
    Object.keys(mockLayoutMatrix).map((curr: string) => {
      //@ts-ignore
      mockLayoutMatrix[curr].forEach((layout: any) => {
        totalRows += layout?.length ||Â 0;
      });
    });
    setTimeout(() => {
      const seats = wrapper.findAll(".seat-rows");
      expect(seats.length).toEqual(totalRows);
    }, 1000);
  });

  test("viewing the allocated seats data should set the right classes", () => {
    const wrapper = mount(App);
    setTimeout(() => {
      expect(wrapper.classes('seats group1 rank1')).toEqual(true);
      const blackSeat = wrapper.findAll(".seats.group1.rank1")[0];
      expect(blackSeat.attributes('stroke')).toBe('black');
      expect(blackSeat.attributes('fill')).toBe('rgb(184, 184, 184)');
    }, 1000);
  });
});
