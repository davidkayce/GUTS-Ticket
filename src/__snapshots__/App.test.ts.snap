// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Theater application should render content correctly 1`] = `
<section
  class="theater"
>
  <section>
    
    
  </section>
  <article>
    <div
      class="card"
    >
      <h3>
        Properties
      </h3>
      <div
        class="card-body"
      >
        <ul
          class="list-group"
        >
          <li
            class="list-group-item"
          >
            <span
              class="group one"
            />
             Group 1
          </li>
          <li
            class="list-group-item"
          >
            <span
              class="group two"
            />
             Group 2
          </li>
          <li
            class="list-group-item"
          >
            <span
              class="group three"
            />
             Group 3 
          </li>
          <li
            class="list-group-item"
          >
            <span
              class="rank one"
            />
             Rank 1
          </li>
          <li
            class="list-group-item"
          >
            <span
              class="rank two"
            />
             Rank 2
          </li>
          <li
            class="list-group-item"
          >
            <span
              class="rank three"
            />
             Rank 3
          </li>
        </ul>
      </div>
    </div>
  </article>
</section>
`;
